backend/tet/events/tests/test_services.py

```diff
@@ -0,0 +1,98 @@
+# not yet designed/implemented, so currently this test only tests for city users, but it needs to be
+# adapted to all users eventually.
+
+# Currently the expectation is that all users getting access to `ServicesClient` have is_staff set.
+
+
+@pytest.mark.django_db
+@override_settings(
+    NEXT_PUBLIC_MOCK_FLAG=False,
+    LINKEDEVENTS_URL='http://localhost/'
+)
+def test_get_postings(requests_mock):
+    """Test that posts are filtered by user's email and divide into published/draft works"""
+    requests_mock.get('http://localhost/event/', json=SAMPLE_EVENTS)
+    user = TetHelsinkiAdUserFactory()
+
+    user.email = "testuser@example.org"
+    postings = ServiceClient().list_job_postings_for_user(user)
+
+    assert len(postings['published']) == 1
+    assert len(postings['draft']) == 1
+
+    user.email = "hasnopostings@example.org"
+    postings = ServiceClient().list_job_postings_for_user(user)
+
+    assert len(postings['published']) == 0
+    assert len(postings['draft']) == 0
+
+    user.email = "otheruser@example.org"
+    postings = ServiceClient().list_job_postings_for_user(user)
+
+    assert len(postings['published']) == 1
+    assert len(postings['draft']) == 0
+
+@pytest.mark.django_db
+@override_settings(
+    NEXT_PUBLIC_MOCK_FLAG=False,
+    LINKEDEVENTS_URL='http://localhost/'
+)
+def test_add_posting(requests_mock):
+    requests_mock.post('http://localhost/event/', json=ADD_EVENT_RESPONSE)
+    event = ServiceClient().add_tet_event(ADD_EVENT_PAYLOAD, TetHelsinkiAdUserFactory())
+    assert event['id'] == 'tet:af7w5v5m6e'
+
+
+@pytest.mark.django_db
+@override_settings(
+    NEXT_PUBLIC_MOCK_FLAG=False,
+    LINKEDEVENTS_URL='http://localhost/'
+)
+def test_edit_tet_posting(requests_mock):
+    matcher = re.compile('.*')
+    # We just care that we get a valid event response back, so ADD_EVENT_RESPONSE is okay to use here
+    requests_mock.put(matcher, json=ADD_EVENT_RESPONSE)
+
+    requests_mock.get('http://localhost/event/tet:af7w4jmjla/', json=EVENT_RESPONSE_TESTUSER)
+    requests_mock.get('http://localhost/event/tet:af7wto6nze/', json=EVENT_RESPONSE_OTHERUSER)
+    requests_mock.get('http://localhost/event/tet:af7wtenvii/', json=EVENT_RESPONSE_NO_USER)
+    requests_mock.get('http://localhost/event/tet:af7wjlxpyy/', json=EVENT_RESPONSE_NO_CUSTOM_DATA)
+
+    # LOGGER.warning(EVENT_RESPONSE_NO_CUSTOM_DATA)
+
+    user = TetHelsinkiAdUserFactory()
+    user.email = "testuser@example.org"
+
+    client = ServiceClient()
+
+    client.update_tet_event('tet:af7w4jmjla', ADD_EVENT_PAYLOAD, user)
+
+    with pytest.raises(PermissionDenied):
+        client.update_tet_event('tet:af7wto6nze', ADD_EVENT_PAYLOAD, user)
+
+    with pytest.raises(PermissionDenied):
+        client.update_tet_event('tet:af7wtenvii', ADD_EVENT_PAYLOAD, user)
+
+    # TODO why doesn't this raise?
+    # with pytest.raises(PermissionDenied):
+    #     client.update_tet_event('tet:af7wjlxpyy', ADD_EVENT_PAYLOAD, user)
```

Comment:
Hmm... this is an interesting case. A test case which does not pass and you have left it in comments.

In this case I would actually say this is ok as this shows the code which you expect to pass but it doesn't.

Other way to do this might be to add this specific test case which fails as a separate parametrization of a test or a separate test altogether and leave it in uncommented and let it fail. Then it would be explicit that we expect this to pass but it does not. Then one might add a comment into the test so anyone who sees the test failing would understand that this is currently normal behavior although not intended.

One downside of adding a failing test case would be that it would complicate code review because one would have to make a habit of mentioning that this specific test case is ok to fail. That might be circumvented at least partially by putting all such TODO/Failing tests into their own file or prefix/suffix the test cases with a string that would make it clear that these test cases are currently failing and it is ok until they are fixed.

I don't see a clear way to handle these kind of cases so for me leaving this as commented out code in this case is fine.

Reply:
Actually after consulting a colleague about this I now know that there's [@pytest.mark.xfail](https://docs.pytest.org/en/6.2.x/skipping.html#xfail-mark-test-functions-as-expected-to-fail) which could be used to mark a test case which is expected to fail. So you could uncomment the failing test code and put it into its own test, mark that with @pytest.mark.xfail and leave a comment explaining the current situation and what's to be done.

Reply:
[Reason ](https://docs.pytest.org/en/6.2.x/skipping.html#reason-parameter) / [raises](https://docs.pytest.org/en/6.2.x/skipping.html#raises-parameter) parameters to @pytest.mark.xfail could be used to add more clarity.

Reply:
Thanks for the information! This was actually something I never intended to make it to the final state of the PR (for some reason git didn't show that "your branch is one commit ahead", so I had one commit I didn't know I hadn't pushed, sorry for that). But good to learn all this.